Prevent opening stopped wallets

lontivero opened the pull request:
Fixes https://github.com/zkSNACKs/WalletWasabi/issues/3408

yahiheb commented on WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:
```diff
@@ -188,7 +188,8 @@ internal WalletViewModelBase OpenWallet(Wallet wallet)
 			{
-				walletViewModel.OpenWalletTabs();
+				if (walletViewModel.Wallet.State != WalletState.Started && walletViewModel.Wallet.State != WalletState.Stopped)
+					walletViewModel.OpenWalletTabs();
```
```suggestion
				{
					walletViewModel.OpenWalletTabs();
				}
```

nopara73 commented:
I don't understand this change. Can you elaborate?

nopara73 commented:
@lontivero I don't understand why you would want to prevent a wallet that's started. Maybe that fixes that specific issue, but it is not good. Also why do you let the wallet to be opened in stopping state? It's really strange for me. Anyhow I made this PR safer in a way that it does not open >= stopping state wallets. Please report back if it fixes the issue.

molnard commented on WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:
```diff
@@ -186,7 +186,7 @@ internal WalletViewModelBase OpenWallet(Wallet wallet)
 			InsertWallet(walletViewModel);
 
-			if (!WalletManager.AnyWallet(x => x.State >= WalletState.Started && x != walletViewModel.Wallet))
+			if (!WalletManager.AnyWallet(x => x.State >= WalletState.Started && x != wallet) && wallet.State < WalletState.Stopping)
```
```suggestion
			if (!WalletManager.AnyWallet(x => x.State == WalletState.Started && x != wallet))
```

lontivero replied:
`wallet.State < WalletState.Stopping` is error-prone and is is not possible to understand what it means, I prefer to be explicit.

nopara73 replied:
@molnard that's not good. Notice `&& wallet.State < WalletState.Stopping` is outside `AnyWallet`.

MaxHillebrand commented:
I encountered a crash of the wallet at loading time, with these logs.
```
<2020-03-31 15:38:35 INFO	WalletManager (202)	Wallet started.
2020-03-31 15:38:36 CRITICAL	Program (49)	System.ArgumentException: An item with the same key has already been added. Key: WalletWasabi.Gui.Controls.WalletExplorer.SendTabViewModel
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at AvalonStudio.Shell.ShellViewModel.AddDocument(IDocumentTabViewModel document, Boolean temporary, Boolean select)
   at AvalonStudio.Shell.IShellExtensions.AddOrSelectDocument[T](IShell me, T document)
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletViewModel.OpenWalletTabs() in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletViewModel.cs:line 106
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.OpenWallet(Wallet wallet) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 191
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.OpenClosedWallet(ClosedWalletViewModel closedWalletViewModel) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 170
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.<.ctor>b__5_0(EventPattern`1 x) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 58
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 54
   at System.Reactive.ObserveOnObserverNew`1.DrainStep(ConcurrentQueue`1 q) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 577
   at System.Reactive.ObserveOnObserverNew`1.DrainShortRunning(IScheduler recursiveScheduler) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 509
   at System.Reactive.ObserveOnObserverNew`1.<>c.<.cctor>b__17_0(IScheduler scheduler, ObserveOnObserverNew`1 self) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 497
   at Avalonia.Threading.AvaloniaScheduler.<>c__DisplayClass2_1`1.<Schedule>b__1()
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority)
   at Avalonia.X11.X11PlatformThreading.HandleX11(CancellationToken cancellationToken)
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode)
   at AvalonStudio.Shell.Shell.StartShellApp[TAppBuilder](TAppBuilder builder, String appName, ShellAppMainDelegate main, String[] args, IFactory layoutFactory)
   at WalletWasabi.Gui.Program.Main(String[] args) in /home/user/WalletWasabi/WalletWasabi.Gui/Program.cs:line 45
2020-03-31 15:38:36 WARNING	Program (85)	System.ArgumentException: An item with the same key has already been added. Key: WalletWasabi.Gui.Controls.WalletExplorer.SendTabViewModel
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at AvalonStudio.Shell.ShellViewModel.AddDocument(IDocumentTabViewModel document, Boolean temporary, Boolean select)
   at AvalonStudio.Shell.IShellExtensions.AddOrSelectDocument[T](IShell me, T document)
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletViewModel.OpenWalletTabs() in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletViewModel.cs:line 106
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.OpenWallet(Wallet wallet) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 191
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.OpenClosedWallet(ClosedWalletViewModel closedWalletViewModel) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 170
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.<.ctor>b__5_0(EventPattern`1 x) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 58
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 54
   at System.Reactive.ObserveOnObserverNew`1.DrainStep(ConcurrentQueue`1 q) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 577
   at System.Reactive.ObserveOnObserverNew`1.DrainShortRunning(IScheduler recursiveScheduler) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 509
   at System.Reactive.ObserveOnObserverNew`1.<>c.<.cctor>b__17_0(IScheduler scheduler, ObserveOnObserverNew`1 self) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 497
   at Avalonia.Threading.AvaloniaScheduler.<>c__DisplayClass2_1`1.<Schedule>b__1()
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority)
   at Avalonia.X11.X11PlatformThreading.HandleX11(CancellationToken cancellationToken)
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode)
   at AvalonStudio.Shell.Shell.StartShellApp[TAppBuilder](TAppBuilder builder, String appName, ShellAppMainDelegate main, String[] args, IFactory layoutFactory)
   at WalletWasabi.Gui.Program.Main(String[] args) in /home/user/WalletWasabi/WalletWasabi.Gui/Program.cs:line 45
Unhandled exception. System.ArgumentException: An item with the same key has already been added. Key: WalletWasabi.Gui.Controls.WalletExplorer.SendTabViewModel
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
   at AvalonStudio.Shell.ShellViewModel.AddDocument(IDocumentTabViewModel document, Boolean temporary, Boolean select)
   at AvalonStudio.Shell.IShellExtensions.AddOrSelectDocument[T](IShell me, T document)
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletViewModel.OpenWalletTabs() in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletViewModel.cs:line 106
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.OpenWallet(Wallet wallet) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 191
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.OpenClosedWallet(ClosedWalletViewModel closedWalletViewModel) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 170
   at WalletWasabi.Gui.Controls.WalletExplorer.WalletExplorerViewModel.<.ctor>b__5_0(EventPattern`1 x) in /home/user/WalletWasabi/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs:line 58
   at System.Reactive.AnonymousSafeObserver`1.OnNext(T value) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 54
   at System.Reactive.ObserveOnObserverNew`1.DrainStep(ConcurrentQueue`1 q) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 577
   at System.Reactive.ObserveOnObserverNew`1.DrainShortRunning(IScheduler recursiveScheduler) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 509
   at System.Reactive.ObserveOnObserverNew`1.<>c.<.cctor>b__17_0(IScheduler scheduler, ObserveOnObserverNew`1 self) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\ScheduledObserver.cs:line 497
   at Avalonia.Threading.AvaloniaScheduler.<>c__DisplayClass2_1`1.<Schedule>b__1()
   at Avalonia.Threading.JobRunner.RunJobs(Nullable`1 priority)
   at Avalonia.X11.X11PlatformThreading.HandleX11(CancellationToken cancellationToken)
   at Avalonia.X11.X11PlatformThreading.RunLoop(CancellationToken cancellationToken)
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken)
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode)
   at AvalonStudio.Shell.Shell.StartShellApp[TAppBuilder](TAppBuilder builder, String appName, ShellAppMainDelegate main, String[] args, IFactory layoutFactory)
   at WalletWasabi.Gui.Program.Main(String[] args) in /home/user/WalletWasabi/WalletWasabi.Gui/Program.cs:line 45
```

and this PR fixes it, I can now load the wallet again.

lontivero commented:
I am sorry team, I messed up this PR until the point that I had no more choice that force a commit. I reverted your commits @nopara73 because they made the wallet crash immediately after pressing "Load Wallet" button.

---------------------

I fixed it in this way because I didn't understand the flow of events. Why is Wasabi trying to open tabs for an already opened wallet when I press Exit?

Basically I saw that the application tries to add the tabs again (Send tab, Receive tab, ....) when I close it and that wallet is in the Stopped or Stopping state, what makes no sense. That's why I added this logic here where I think is the best place.

I think there is a deeper issue here but it can be much harder to find and fix, and that could create new problems too.

danwalmsley commented:
My diagnosis of the issue:

1) we subscribe to wallet state changes here: https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs#L46

2) that subscription is observed on RxApp.MainThreadScheduler

3) when the handler runs eventArgs is now out of date, and it will say WalletState == Started, however in reality the wallet has now stopped.

The correct fix:
change: https://github.com/zkSNACKs/WalletWasabi/blob/master/WalletWasabi.Gui/Controls/WalletExplorer/WalletExplorerViewModel.cs#L56

to say:
```
else if (_walletDictionary[wallet] is ClosedWalletViewModel cwvm && wallet.State == WalletState.Started)
```

danwalmsley commented:
@lontivero iv pushed the fix where the check is in the original subscription so we dont have 2 places where we make the check.

lontivero commented:
tACK

lontivero commented:
Dan and I were okay with this change.
